src/graph.c

```diff
@@ -221,4 +223,206 @@ float graph_distance(const Graph* graph1, const Graph* graph2)
+  
+    // destroy bitsets used in past calculations
+    destroy_bitset(newP);
+    destroy_bitset(newR);
+    destroy_bitset(newX);
+  
+    // P := P \ {v}
+    unset_bit(P, v);
+    // X := X ⋃ {v}
+    set_bit(X, v);
+  }
+}
+
+Bitset** construct_bitset_adjacency_matrix(Graph *graph) {
+  ASSERT(graph != NULL, "graph is NULL");
+  Bitset **bitset_adjacency_matrix = (Bitset**)calloc(graph->vertices + 1, sizeof(Bitset*));
+  for (int v = 1; v <= graph->vertices; ++v) {
+    bitset_adjacency_matrix[v] = create_bitset(graph->vertices + 1);
+  }
+  for (int v = 1; v <= graph->vertices; ++v) {
+    for (int u = v + 1; u <= graph->vertices; ++u) {
+      if (graph->adjacency_matrix[v][u] && graph->adjacency_matrix[u][v]) {
+        set_bit(bitset_adjacency_matrix[v], u);
+        set_bit(bitset_adjacency_matrix[u], v);
+      }
+    }
+  }
+  return bitset_adjacency_matrix;
+}
+
+void destroy_bitset_adjacency_matrix(Bitset** bitset_adjacency_matrix, int vertices) {
+  for (int i = 1; i <= vertices; ++i) {
+    destroy_bitset(bitset_adjacency_matrix[i]);
+  }
+  free(bitset_adjacency_matrix);
+}
+
+Graph* extract_clique(Graph *graph, Bitset *clique) {
+  ASSERT(graph != NULL, "graph is NULL");
+  ASSERT(graph != NULL, "clique is NULL");
+  Graph *extracted_clique = (Graph*)calloc(1, sizeof(Graph));
+  initialize_graph(extracted_clique, graph->vertices, 0);
+  for (int v = 1; v <= graph->vertices; ++v) {
+    if (!get_bit(clique, v)) continue;
+    for (int u = v + 1; u <= graph->vertices; ++u) {
+      if (!get_bit(clique, u)) continue;
+      if (graph->adjacency_matrix[v][u]) {
+        extracted_clique->adjacency_matrix[v][u] = graph->adjacency_matrix[v][u];
+        extracted_clique->edges += graph->adjacency_matrix[v][u];
+      }
+      if (graph->adjacency_matrix[u][v]) {
+        extracted_clique->adjacency_matrix[u][v] = graph->adjacency_matrix[u][v];
+        extracted_clique->edges += graph->adjacency_matrix[u][v];
+      }
+      if (extracted_clique->adjacency_matrix[v][u] || extracted_clique->adjacency_matrix[u][v]) {
+        extracted_clique->unique_edges++;
+      }
+    }
+  }
+  return extracted_clique;
+}
+
+uint32_t clique_get_max_p(Graph *clique) {
+  ASSERT(clique != NULL, "clique is NULL");
+  uint32_t p = clique->vertices * clique->vertices;
+  for (int v = 1; v <= clique->vertices; ++v) {
+    for (int u = v + 1; u <= clique->vertices; ++u) {
+      if (!clique->adjacency_matrix[v][u] || !clique->adjacency_matrix[u][v]) continue;
+      p = fmin(p, fmin(clique->adjacency_matrix[v][u], clique->adjacency_matrix[u][v]));
+    }
+  }
+  return p;
+}
+
+uint8_t p_clique_cmp(Graph *clique1, uint32_t clique1_size, uint32_t p1, Graph *clique2, uint32_t clique2_size, uint32_t p2) {
+  ASSERT(clique1 != NULL, "clique1 is NULL");
+  ASSERT(clique2 != NULL, "clique2 is NULL");
+  if (clique1_size > clique2_size) return -1;
+  if (clique1_size < clique2_size) return 1;
```

Comment:
Chyba 2 ify z clique_size mozna wywalic bo usuwamy w BronKerboschu mniejsze

Reply:
Racja. Zapomniałem to usunąc po optymalizacji bufora klik i dodaniu store_clique(). Wywaliłem size z porównania i obliczanie tablicy size z get_max_clique().